synthesizer/benches/deployment.rs

```diff
@@ -0,0 +1,174 @@
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg(feature = "testing")]
+
+#[macro_use]
+extern crate criterion;
+
+mod benchmarks;
+use benchmarks::*;
+
+mod utilities;
+use utilities::*;
+
+use console::network::Testnet3;
+use snarkvm_synthesizer::{ConsensusStorage, Transaction, VM};
+
+use criterion::{BatchSize, Criterion};
+use std::fmt::Display;
+
+const NUM_DEPLOYMENTS: &[usize] = &[2, 4, 8, 16, 32, 64, 128, 256];
+// Note: The maximum number of mappings that can be included in a program is 31.
+const NUM_MAPPINGS: &[usize] = &[1, 2, 4, 8, 16, 31];
+
+/// A helper function to remove programs, from a set of deployment transactions, from the finalize store.
+fn clean_finalize_state<C: ConsensusStorage<Testnet3>>(vm: &VM<Testnet3, C>, transactions: &[Transaction<Testnet3>]) {
+    for transaction in transactions {
+        match transaction {
+            Transaction::Deploy(_, _, deployment, _) => {
+                let program = deployment.program();
+                if vm.finalize_store().contains_program_confirmed(program.id()).unwrap() {
+                    for (mapping_name, _) in program.mappings() {
+                        vm.finalize_store().remove_mapping(program.id(), mapping_name).unwrap();
+                        vm.process().write().remove_program(program.id()).unwrap();
+                    }
+                    vm.finalize_store().remove_program(program.id()).unwrap();
+                }
+            }
+            _ => unreachable!("The benchmark should only contain deployment transactions."),
+        }
+    }
+}
+
+/// A helper function for benchmarking deployments.
+pub fn bench_deployment<C: ConsensusStorage<Testnet3>>(
+    c: &mut Criterion,
+    header: impl Display,
+    mut workload: Workload,
+) {
+    // Setup the workload.
+    let (vm, private_key, benchmark_transactions, mut rng) = workload.setup::<C>();
+
+    // Run the benchmarks
+    for (name, transactions) in benchmark_transactions {
+        if transactions.is_empty() {
+            println!("Skipping benchmark {} because it has no transactions.", name);
+            continue;
+        }
+
+        let mut num_transactions = 0f64;
+        let mut num_rejected = 0f64;
+
+        // Benchmark `speculate`.
+        c.bench_function(&format!("{header}/{name}/speculate"), |b| {
+            b.iter_batched(|| {}, |_| vm.speculate(transactions.iter()).unwrap(), BatchSize::SmallInput)
+        });
+
+        // Construct the `Transactions` object.
+        let txns = vm.speculate(transactions.iter()).unwrap();
+
+        // Benchmark `finalize`.
+        c.bench_function(&format!("{header}/{name}/finalize"), |b| {
+            b.iter_batched(
+                || {
+                    vm.finalize_store().start_atomic();
+                    clean_finalize_state(&vm, &transactions);
+                    vm.finalize_store().finish_atomic().unwrap();
```

Comment:
it might make sense to put `start_atomic` and `finish_atomic` inside `clean_finalize_state`